packages/runtime/container-runtime/src/dataStore.ts

```diff
@@ -0,0 +1,151 @@
+    /** The alias name to be assigned to the datastore */
+    readonly alias: string;
+}
+
+/**
+ * Type guard that returns true if the given alias message is actually an instance of
+ * a class which implements @see IDataStoreAliasMessage
+ * @param maybeDataStoreAliasMessage - message object to be validated
+ * @returns True if the @see IDataStoreAliasMessage is fully implemented, false otherwise
+ */
+ export const isDataStoreAliasMessage = (
+    maybeDataStoreAliasMessage: any,
+): maybeDataStoreAliasMessage is IDataStoreAliasMessage => {
+    return typeof maybeDataStoreAliasMessage?.internalId === "string"
+        && typeof maybeDataStoreAliasMessage?.alias === "string";
+};
+
+/**
+ * A fluid router with the capability of being assigned an alias
+ */
+ export interface IDataStore extends IFluidRouter {
+    /**
+     * Attempt to assign an alias to the datastore.
+     * If the operation succeeds, the datastore can be referenced
+     * by the supplied alias.
+     *
+     * @param alias - Given alias for this datastore.
+     */
+    trySetAlias(alias: string): Promise<boolean>;
+}
+
+export const channelToDataStore = (
+    fluidDataStoreChannel: IFluidDataStoreChannel,
+    internalId: string,
+    runtime: ContainerRuntime,
+    logger: ITelemetryLogger,
+): IDataStore => new DataStore(fluidDataStoreChannel, internalId, runtime, logger);
+
+enum AliasState {
+    Aliased = "Aliased",
+    Aliasing = "Aliasing",
+    None = "None",
+}
+
+class DataStore implements IDataStore {
+    private aliasState: AliasState = AliasState.None;
+    private alias: string | undefined;
+
+    async trySetAlias(alias: string): Promise<boolean> {
+        assert(this.runtime.attachState === AttachState.Attached, "Trying to alias when not attached!");
+        switch (this.aliasState) {
+            // If we're already aliasing, fail the current function call
+            case AliasState.Aliasing: return false;
+            // If this datastore is already aliased, return true only if this
+            // is a repeated call for the same alias
+            case AliasState.Aliased: return this.alias === alias;
+            // There is no current or past alias operation for this datastore,
+            // it is safe to continue execution
+            case AliasState.None: break;
+            default: unreachableCase(this.aliasState);
+        }
+
+        this.aliasState = AliasState.Aliasing;
+        this.fluidDataStoreChannel.bindToContext();
+
+        const message: IDataStoreAliasMessage = {
+            internalId: this.internalId,
+            alias,
+        };
+
+        return this.ackBasedPromise<boolean>((resolve) => {
+            this.runtime.submitDataStoreAliasOp(message, resolve);
+        }).then((succeeded) => {
+            if (succeeded) {
+                this.aliasState = AliasState.Aliased;
+                this.alias = alias;
+            } else {
+                this.aliasState = AliasState.None;
```

Comment:
None means one can retry (with same or different name). While there is nothing wrong with our logic to allow retry (using different name), I'm not sure I see a scenario for that. I feel like all scenarios we know are centered around using well-defined and known name upfront, and if it failed, it's pretty much game over (for this object, from app perspective).
I'd probably restrict API usage to only that scenario - we can always remove these limitations if someone proves to us they have a reason to extend API outreach.